Evaluation metrics

mohit-surana opened the pull request:
- Hits@K
- AP@K
- MAP@K

mohit-surana committed:
Evaluation metrics

mudit1990 commented on code/metrics.py:
```diff
@@ -0,0 +1,108 @@
+import numpy as np
+
+
+def hits_at_k(y_true, y_pred=None, k=None, is_l2=False):
```
I think hits@k should be defined from both ends. We should know how well we are doing in finding if mails are written by sender, recv and how well are predicting suspicious mails

mudit1990 commented on code/metrics.py:
```diff
@@ -0,0 +1,108 @@
+            correct += y_true[i]
+        return correct, k
+    else:
+        # This does not have clear semantics if we are given the binary probabilities
```
this is not very clear, shouldn't binary probabilities follow the same definition above

mohit-surana replied:
I think you're right.
I'll change this.

mudit1990 commented on code/metrics.py:
```diff
@@ -0,0 +1,108 @@
+        correct = 0
+        for i in xrange(k):
+            correct += y_true[i]
+        return correct, k
```
shouldn't we return a ratio e.g correct/k ?

mohit-surana replied:
I believe both are equivalent but it seemed to me that we will not be saying that we are getting .8 hits @ K; instead say 4 hits @ 5

mudit1990 commented on code/metrics.py:
```diff
@@ -0,0 +1,108 @@
+        correct += y_true[i]
+        precision = correct / (i+1.0)
+        ap += precision * y_true[i]
+    return ap / np.sum(y_true), k
```
shouldn't this be ap/k
Think about this, if we get everything right and k is 3 and y_true has 10 elements out of which 6 are 1. You AP@k will be 3/6 = 0.5, while it should have been 1
Reference: https://github.com/benhamner/Metrics/blob/master/Python/ml_metrics/average_precision.py

mohit-surana replied:
Different people use different notations. This is partly why it took a long time to write this because different views contradicted each other.

This is what I followed:
https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Average_precision

For the example you are describing, we cannot say much unless we know the relative ordering of the elements
Say it is
a) 1000011111, k=3
Looking at the first three docs, we get (1/1 * 1 + 1/2 * 0 + 1/3 * 0) and then should do a by 3
b) 1100001111, k=3
Looking at the first three docs, we get (1/1 * 1 + 2/2 * 1 + 2/3 * 0) and then should do a by 3

Now the /3 does not come because of the just the value of k
There were other documents that should have been selected preferentially.

That being said, I think there still is a small mistake
It should be min(k, np.sum(y_true))

mudit1990 replied:
which is k by ur definition in line 43 to 46

mohit-surana replied:
There is a subtle difference.
min(k, len(y_true)) is to handle asking for precision@20 when we have only 10 documents

min(k, np.sum(y_true)) is so that we don't penalize a person if they give us all the relevant documents
Eg. there were 4 relevant documents, we asked for 10

mohit-surana committed:
Evaluation metrics
- bug fixes and refactor
- included results for both positive and negative queries

mohit-surana committed:
Evaluation metrics - updated documentation

mohit-surana commented:
Merging for now. 
Review comments will be acted upon via subsequent commits
